Feature/code formatter

RayGuo-ergou opened the pull request:
# Description

Formatter for c/c++

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Tested on newest version of Ubuntu.

Only one file has been formatted and push is  _code/c/control-flow/rocket_launch.c_, because other people's migration PR have not been approved or created. Therefore, to avoid the conflict I just format one file as example.

# Checklist:

- [x] I have performed a self-review of my own code

RayGuo-ergou committed:
feat: code formatter with an example

RayGuo-ergou committed:
docs: add readme file including code formater

RayGuo-ergou committed:
docs: only format c and cpp file in usage section

ghost commented:
Great initiative @RayGuo-ergou ! I will leave the reviewing of the code to @tancnle.
But I am curious, does this format on save, upload, or is it ran manually?

RayGuo-ergou commented:
@talzid Thanks for the comment, if the clang-format VS code extension installed and selected as the default code formatter. Then you can use the format on save function on VS code. Otherwise, should use the command to manually format the code as I introduced in the README file.

ghost commented:
> @talzid Thanks for the comment, if the clang-format VS code extension installed and selected as the default code formatter. Then you can use the format on save function on VS code. Otherwise, should use the command to manually format the code as I introduced in the README file.

Excellent point, you did a great job of explaining it in the readme.md. This is a lesson to myself in why not to skim read 😄 
Personally, I think everyone should use `npm` , solely as it _should_ be installed by the majority of Thoth Tech for documentation purposes as well.

tancnle commented on README.md:
```diff
@@ -0,0 +1,135 @@
+
+We use [_clang-format_](https://clang.llvm.org/docs/ClangFormat.html) as our code format for this project.
+
+## installation
```
```suggestion
## Installation
```

RayGuo-ergou replied:
Thanks for the suggestion will do.

tancnle commented on .clang-format:
```diff
@@ -0,0 +1,192 @@
+---
+Language:        Cpp
+# BasedOnStyle:  LLVM
```
**question:** Is this comment needed? Are these rules a copy from an available [set](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#configurable-format-style-options)? If yes, can we use the predefined one? 🤔 Google [style](https://google.github.io/styleguide/cppguide.html) is a good candidate but I could be biased.

```yaml
---
BasedOnStyle: Google
```

RayGuo-ergou replied:
Actually this is my first time working with c++ so I just picked the basic style. But I think you have a lot experience with c/c++, so I think we can change to google style.

tancnle replied:
> But I think you have a lot experience with c/c++, so I think we can change to google style.

I don't really have much exp. to be honest with you 😅 I just noticed that the existing styles are more or less similar to the Google style so hence I suggest that. In addition, I think the Google style [documentation](https://google.github.io/styleguide/cppguide.html) is quite good and easy to follow.

tancnle replied:
@RayGuo-ergou can we just specify this in the `.clang-format` file?

```yaml
---
BasedOnStyle: Google
```

RayGuo-ergou replied:
Yes, will try this

RayGuo-ergou replied:
> > But I think you have a lot experience with c/c++, so I think we can change to google style.
> 
> I don't really have much exp. to be honest with you 😅 I just noticed that the existing styles are more or less similar to the Google style so hence I suggest that. In addition, I think the Google style [documentation](https://google.github.io/styleguide/cppguide.html) is quite good and easy to follow.

Right the good document can always be a result why we use it. So I think we can keep google style for now.

tancnle commented on .clang-format:
```diff
@@ -0,0 +1,192 @@
+StatementMacros:
+  - Q_UNUSED
+  - QT_REQUIRE_VERSION
+TabWidth:        8
```
**nitpick (non-blocking):** Could we remove extraneous spaces between the colon `:` and rule values across this file 🙏🏼 ?

RayGuo-ergou replied:
yeah of course my bad forgot to format this yml file.

tancnle commented on README.md:
````diff
@@ -0,0 +1,135 @@
+In order to recursively format all files, you can use command:
+
+```bash
+find . -regex '.*\.\(cpp\|c\)' -exec clang-format -style=file -i {} \;
````
**suggestion (non-blocking):** We could put this command in a `Makefile`, something like `make format`.

RayGuo-ergou replied:
Haven't done any work with `Makefile` but I can research and start on that, thanks for the suggestion.

tancnle commented on code/c/control-flow/rocket_launch.c:
```diff
@@ -1,8 +1,8 @@
-#include <stdio.h>
-#include <stdbool.h>
 #include "splashkit.h"
+#include <stdbool.h>
```
**suggestion:** Alphabetical ordering is ok. However, I think it is better to put the C standard library before the project `.h` files (see Google [style](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)). What do you think?

tancnle commented on README.md:
````diff
@@ -0,0 +1,135 @@
+sudo apt install clang-format
+```
+
+### MacOS
````
```suggestion
### macOS
```

tancnle reviewed (commented):
Thank you for your hard work, @RayGuo-ergou ❤️ This is great. I have added some suggestions for your consideration. Please let me know what you think 🙏🏼 .

Have you thought about integrating this into the CI? Once we fix all the violations, I think we can run something like `clang-format --dry-run code/c/control-flow/test-if.c` as part of the build to automatically detect issue.

<details>

```shell
❯ clang-format --dry-run code/c/control-flow/test-if.c
code/c/control-flow/test-if.c:5:11: warning: code should be clang-formatted [-Wclang-format-violations]
int main()
          ^
code/c/control-flow/test-if.c:6:2: warning: code should be clang-formatted [-Wclang-format-violations]
{
 ^
code/c/control-flow/test-if.c:7:19: warning: code should be clang-formatted [-Wclang-format-violations]
    int num, num1;
                  ^
code/c/control-flow/test-if.c:9:32: warning: code should be clang-formatted [-Wclang-format-violations]
    printf("Enter a number: ");
                               ^
code/c/control-flow/test-if.c:10:23: warning: code should be clang-formatted [-Wclang-format-violations]
    scanf("%d", &num);
                      ^
code/c/control-flow/test-if.c:12:18: warning: code should be clang-formatted [-Wclang-format-violations]
    if (num != 2)
                 ^
code/c/control-flow/test-if.c:13:35: warning: code should be clang-formatted [-Wclang-format-violations]
        printf("Num is not 2!\n");
                                  ^
code/c/control-flow/test-if.c:15:38: warning: code should be clang-formatted [-Wclang-format-violations]
    printf("Enter another number: ");
                                     ^
code/c/control-flow/test-if.c:16:24: warning: code should be clang-formatted [-Wclang-format-violations]
    scanf("%d", &num1);
                       ^
code/c/control-flow/test-if.c:18:31: warning: code should be clang-formatted [-Wclang-format-violations]
    if (num1 == 2 && num != 2)
                              ^
code/c/control-flow/test-if.c:19:50: warning: code should be clang-formatted [-Wclang-format-violations]
        printf("You got the hint... num1 is 2!");
                                                 ^
code/c/control-flow/test-if.c:22:20: warning: code should be clang-formatted [-Wclang-format-violations]
    if (num > num1)
                   ^
code/c/control-flow/test-if.c:23:64: warning: code should be clang-formatted [-Wclang-format-violations]
        printf("The first number you entered was the larger.");
                                                               ^
code/c/control-flow/test-if.c:24:9: warning: code should be clang-formatted [-Wclang-format-violations]
    else
        ^
code/c/control-flow/test-if.c:25:64: warning: code should be clang-formatted [-Wclang-format-violations]
        printf("The first number you entered was not larger.");
                                                               ^
```

<details>

RayGuo-ergou committed:
docs: Change format style to google, add script for CI and fix typo in
readme file

RayGuo-ergou committed:
feat: add github action to check the format

RayGuo-ergou committed:
ci: change ci name to check code format

RayGuo-ergou committed:
feat: add a make file to format all code in the project

RayGuo-ergou committed:
refactor: add a newline in the end of makefile

RayGuo-ergou commented:
![image](https://user-images.githubusercontent.com/33137074/167175157-dde25c53-ffdc-427a-a105-ecccbd7ed2b6.png)
![image](https://user-images.githubusercontent.com/33137074/167175168-31f32ebb-f47e-4ada-b8e5-fba8b5f64844.png)

Here is the test for the `Makefile`, as I mentioned in the PR Description to avoid the conflict so I have not added and push the formatted file on github

RayGuo-ergou commented on README.md:
````diff
@@ -0,0 +1,177 @@
+find . -regex '.*\.\(cpp\|c\)' -exec clang-format -style=file -i {} \;
+```
+
+## Check files
````
@tancnle Any suggestion for this section title? I do not think this title is appropriate.

talzid replied:
Ideally, it should state what that section is about. I suggest being more specific. Changing _check_ to format or updating formatting or _validate code_

RayGuo-ergou replied:
Thanks for the suggestion

tancnle replied:
@RayGuo-ergou Clang-format is a code formatter, it highlights code style issues. I would vote for `Format code` to keep the language consistent. `Validate` could possibly be inferred as testing as well.

RayGuo-ergou commented:
Hi @tancnle I have addressed the update based on your suggestion.

RayGuo-ergou committed:
refactor: update readme for makefile and use basedonstyle instead of all options for format config file

tancnle commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,23 @@
+          sudo apt update
+          sudo apt install -y clang-format
+      - name: checkout repository content
+        uses: actions/checkout@v2
```
**suggestion:** I think the latest version is [v3](https://github.com/actions/checkout) at the moment.

```suggestion
        uses: actions/checkout@v3
```

tancnle commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,23 @@
+      - name: setup python
+        uses: actions/setup-python@v3
+      - name: run python Script
+        run: |
+          python3 run-clang-format.py -r .
```
**nitpick (non-blocking):** This can be collapsed to one-line 🤔 

```suggestion
        run: python3 run-clang-format.py -r .
```

tancnle commented on run-clang-format.py:
```diff
@@ -0,0 +1,408 @@
+#!/usr/bin/env python
```
**suggestion:** Can we move this file under the `scripts` folder and make it executable? It helps to declutter the root directory.

```shell
chmod +x run-clang-format.py
```

tancnle commented on run-clang-format.py:
```diff
@@ -0,0 +1,408 @@
+#!/usr/bin/env python
```
**suggestion:** The wrapper script adds some really conveniences on top of the default `clang-format` command.

- Parallel execution. The sequential run is still pretty fast given the small codebase.
- Code diff output.

On the other hand, this adds a new dependency to the project, that is `python3`. While most modern OS comes bundled with Python 3, Windows might require additional setup.

Weighting between cost vs benefits, I wonder if there is the right time to introduce the wrapper script 🤔. I could have missed the rationale behind this change though. Keen to hear your thoughts.

RayGuo-ergou replied:
I was actually thinking the same thing, linux ( maybe mac, have no experience with mac so not sure) has python3 as default package, and to install python on windows is pretty straight forward. I did consider the windows users who have not installed python. That's why I use the ` find . regex .... ` command in the makefile and Readme. But I think we can keep the python script because I think it easlier to use than ` clang-format ` command(e.g. simply use option ` -r ` can go through every files and color indication). We can have both command (script and `clang-format`) introduce in the readme file. Therefore, its not mandatory for windows users install python.

tancnle commented on Makefile:
```diff
@@ -0,0 +1,2 @@
+format: 
```
**thought**: `make` is readily available in most OSes. Windows will require WSL2 to be installed.

RayGuo-ergou replied:
will add this info in readme file
update: will wait for your reply to modify the readme file

tancnle commented on README.md:
````diff
@@ -0,0 +1,177 @@
+In order to recursively format all files, you can use command:
+
+```bash
+make format
````
**question:** Can the wrapper script format the code as well? I think we should either go with `make` wrapper for the clang-format command or the wrapper script (see more about my [thought](https://github.com/thoth-tech/programming-arcana/pull/2/files#r867419985) to stay with the earlier).

RayGuo-ergou replied:
sure thing

RayGuo-ergou replied:
Yes we can use wrapper script to format code.From my opinoin we can use the wrapper script and remove the `makefile`. Then intoduce the command ` find . -regex '.*\.\(cpp\|c\)' -exec clang-format -style=file -i {} \;` in readme for windows users as my [reply](https://github.com/thoth-tech/programming-arcana/pull/2#discussion_r867529693) because windows users need wsl to running the ` make ` command. So I think there's no point to have a ` makefile ` with 
```
format: 
	find . -regex '.*\.\(cpp\|c\)' -exec clang-format -style=file -i {} \;
```

tancnle replied:
Is `find` available in Windows or do we still need WSL 🤔 ?

How about a command without `find`?

```
clang-format -style=file -i **/*.(cpp|c)
```

RayGuo-ergou replied:
Ops forget the ` find ` command is not available in windows, I will try to use powershell command instead.

And unfortunate clang-format does not support recurse function.

RayGuo-ergou committed:
refactor: change action CI and wrapped script

Change the Action CI to use --dry-run option which is list file names
instead of list all code, change the exit code for --dry-run from 0 to 1
so can give error if file required format.

RayGuo-ergou committed:
refactor: change CI and wrapped script

missing understand the --dry-run option(list all files not files
required format) change to -quite just for the
exit code 1

tancnle commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,24 @@
+        uses: actions/setup-python@v3
+      - name: run python Script
+        run: |
+          chmod +x ./script/run-clang-format.py
```
**suggestion:** We should commit this mod in git and remove this line. What do you think?

RayGuo-ergou replied:
done with command ` git update-index --chmod=+x ./script/run-clang-format.py  `

RayGuo-ergou committed:
refactor: change wrapper script to executable and remove 'chmod' in ci

RayGuo-ergou committed:
docs: remove makefile, add wrapped script and powershell command

RayGuo-ergou commented:
Hi @tancnle updated the document, the windows users can use the command in document without install python3.

tancnle commented on README.md:
````diff
@@ -0,0 +1,192 @@
+For example, you can use check if the file _./code/c/array/array-copy.c_ needs to be formatted.
+
+```bash
+# make the wrapped script executable
+chmod +x ./script/run-clang-format.py
````
**question:** Is this still needed given the file is already executable?

```shell
stat -c %A script/run-clang-format.py
-rwxr-xr-x
```

RayGuo-ergou replied:
Right, we don't need to. forget we changed the mod in git.

tancnle commented on README.md:
````diff
@@ -0,0 +1,192 @@
+
+The output would be
+
+```
````
**nitpick (non-blocking)**: 

```suggestion
```diff
```

tancnle commented on README.md:
````diff
@@ -0,0 +1,192 @@
+ }%
+```
+
+\*Python3 is normally installed as default package for most operation systems. However, if you are using windows and do not want to install Python. You can use:
+`clang-format .\code\c\array\array-copy.c `, but it would not have the feature in the wrapped script e.g. color.
````
```suggestion
\*Python3 is normally installed as a default package in most operating systems. However, if you are using Windows and do not want to install Python, you can use
`clang-format .\code\c\array\array-copy.c `. However,  it would not have the feature in the wrapper script, for example, coloured output.
```

tancnle commented on README.md:
```diff
@@ -0,0 +1,192 @@
+\*Python3 is normally installed as default package for most operation systems. However, if you are using windows and do not want to install Python. You can use:
+`clang-format .\code\c\array\array-copy.c `, but it would not have the feature in the wrapped script e.g. color.
+
+### Replace the Old Code with Formatted Code
```
**suggestion:** According to our documentation [guide](https://github.com/thoth-tech/handbook/blob/main/docs/processes/documentation/writing-style-guide.md#capitalisation), we should use sentence case

```suggestion
### Replace the old code with formatted code
```

or simply

```suggestion
### Format code
```

RayGuo-ergou commented:
@tancnle Thanks for your feedback will change accroding to that

RayGuo-ergou committed:
docs: update readme file to fix grammar and word use

tancnle commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,22 @@
+        uses: actions/checkout@v3
+      - name: setup python
+        uses: actions/setup-python@v3
+      - name: run python Script
```
**suggestion:** We could be more specific 🤔 

```suggestion
      - name: Clang format
```

RayGuo-ergou committed:
CI: change the name of checking format

change the name from running python script to clang format check for github action

tancnle commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,22 @@
+      - name: setup python
+        uses: actions/setup-python@v3
+      - name: run python Script
+        run: ./script/run-clang-format.py -rq .
```
**suggestion:** We must keep the CI green at all times (ideally) :) Could we mark this job `continue-on-error: true` so the failure is skipped and does not reflect on the overall pipeline result 🤔 ?

RayGuo-ergou replied:
> **suggestion:** We must keep the CI green at all times (ideally) :) Could we mark this job `continue-on-error: true` so the failure is skipped and does not reflect on the overall pipeline result 🤔 ?

Yes we can, but I think it's better to keep the current approach. The CI will pass (green) if all files follow the format config.

Or you want to have this approach: CI passes anyway but output the files/code need to be formatted?

tancnle replied:
> CI passes anyway but output the files/code need to be formatted?

Yep. We can do a follow-up PR to format all the files and remove this flag.

RayGuo-ergou replied:
> Yep. We can do a follow-up PR to format all the files and remove this flag.

Sure thing will do.

RayGuo-ergou committed:
CI: make CI always pass and output the code that need format.

tancnle commented on .github/workflows/CI.yml:
```diff
@@ -19,4 +19,4 @@ jobs:
         uses: actions/setup-python@v3
       - name: clang format check
-        run: ./script/run-clang-format.py -rq .
+        run: ./script/run-clang-format.py -r .
```
**question:** I am curious of why we do not run this in `quiet` mode (`-q`)? The verbose output s hard to read on the CI iand might be more suitable for running in the local environment. WDYT?

RayGuo-ergou replied:
The ` quiet ` mode would not have any output whether there are files need to be formatted or not. I thought we want some Indicates like which file should be formated.

BTW, there are such funciton in the wrapped script but I can add it. What do you think we just output the filename that needs to be formatted?

tancnle commented on script/run-clang-format.py:
```diff
@@ -397,8 +397,6 @@ def main():
             if not args.quiet:
                 print_diff(outs, use_color=colored_stdout)
-            if retcode == ExitStatus.SUCCESS:
-                retcode = ExitStatus.DIFF
```
**question:** Would this mean the exit code no longer reflect the correct state of the formatter (ie. format issues do not return a non-zero exit code)? If yes, I would propose not to modify the behaviour of the original script.

Few ideas,

1. Mark the step as `continue-on-error: true`
  ```yml
   - name: clang format check
        run: ./script/run-clang-format.py -rq .
        continue-on-error: true
   ```
2. Remove the CI step altogether. We will add it with all the format fixes in a separate PR.

I am leaning towards Option 1. WDYT?

RayGuo-ergou replied:
I think I misunderstood the word "flag", I thought the flag is the error flag for CI, just confirm is it the ` continue-on-error: true`? So like we can have this flag to make the CI pass then remove this in a follow up PR to format all code?

I perfer option1 as well.

tancnle replied:
> I think I misunderstood the word "flag", I thought the flag is the error flag for CI

Sorry for the confusion 🙇🏼 

> just confirm is it the `continue-on-error: true`? So like we can have this flag to make the CI pass then remove this in a follow up PR to format all code?

Yep.

RayGuo-ergou committed:
CI: give flag continue-on-error so the ci always passes

will move the flag in a follow up PR to format all the codes

RayGuo-ergou commented:
Tested with `./script/run-clang-format.py -lr . | grep "rocket"`

Output: 
```bash
./topics/control-flow/examples/rocket_launch1.cpp
./topics/control-flow/examples/rocket_launch2.cpp
./topics/control-flow/examples/rocket_launch.cpp
```

Which dose not include the formatted file ` code/c/control-flow/rocket_launch.c `

RayGuo-ergou commented:
misclicked Close with comment..

tancnle reviewed (approved):
Thanks, @RayGuo-ergou 🙇🏼‍♂️ The MR looks great to me. Let's ship it 👍🏼 🚀
